design/metadata-handling.md

```diff
@@ -0,0 +1,668 @@
+- **gateway**: the gateway to use
+- **services**: a list of services object as defined later
+
+The **networks/ipv4Dhcp** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+The **networks/ipv6** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **ipAddress**: the IP address object
+- **netmask**: the netmask, in an integer format
+- **routes**: the list of route objects
+
+The **networks/ipv6Dhcp** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+The **networks/ipv6Slaac** object contains the following:
+
+- **id**: the network name
+- **link**: The name of the link to configure this network for
+- **routes**: the list of route objects
+
+##### the services specifications
+
+The object for the **services** section list can be one of:
+
+- **dns**: a dns service with the ip address of a dns server
+
+### The Metal3Data object
+
+The output of the controller would be a Metal3Data object,one per node linking to the
+Metal3DataTemplate object and the associated secrets
+
+The Metal3Data object would be:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Data
+metadata:
+  name: nodepool-1-0
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3DataTemplate
+    name: nodepool-1
+spec:
+  index: 0
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-metadata
+    namespace: default
+  metal3Machine:
+    name: machine-1
+    namespace: default
+```
+The Metal3Data will contain the index of this node, and links to the secrets
+generated and to the Metal3Machine using this Metal3Data object.
+
+If the Metal3DataTemplate object is updated, the generated secrets will not be
+updated, to allow for reprovisioning of the nodes in the exact same state as
+they were initially provisioned. Hence, to do an update, it is necessary to do
+a rolling upgrade of all nodes.
+
+The reconciliation of the Metal3DataTemplate object will also be triggered by
+changes on Metal3Machines. In the case that a Metal3Machine gets modified, if
+the `dataTemplate` references a Metal3DataTemplate, that object will be reconciled.
+There will be two cases:
+
+- An already generated Metal3Data object exists with an ownerReference to this
+  Metal3Machine. In that case, the reconciler will verify that the required
+  secrets exist. If they do not, they will be created.
+- if no secret exists with an ownerReference to this Metal3Machine, then the
+  reconciler will create one and fill the respective field with the secret name.
+
+To create a Metal3Data object, the controller will select an index for that
```

Comment:
Let's be specific when we say "the controller" -- which controller? The one for the template CRD?

Reply:
the Metal3DataTemplate controller, thanks